File: lib/core/services/api_client.dart
import 'package:dio/dio.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';

import '../constants/backend_config.dart';
import '../errors/app_errors.dart';
import 'request_log.dart';
export '../errors/app_errors.dart';

/// The sole HTTP entry point. Keep API keys on the server, never in this app.
class ApiClient {
  ApiClient._() {
    final resolved = resolveBackendUrl(
      dartDefineUrl: const String.fromEnvironment('BACKEND_URL'),
      dotenvUrl: dotenv.isInitialized ? dotenv.env['BACKEND_URL'] : null,
    );
    _dio = Dio(BaseOptions(
      baseUrl: resolved,
      connectTimeout: const Duration(seconds: 20),
      receiveTimeout: const Duration(seconds: 60),
      headers: {'Accept': 'application/json'},
      validateStatus: (s) => s != null && s < 500,
    ));
    _dio.interceptors.add(InterceptorsWrapper(onRequest: (options, handler) {
      options.headers['Accept-Language'] = _language;
      handler.next(options);
    }));
  }
  static final ApiClient instance = ApiClient._();
  late final Dio _dio;
  String _language = 'en';
  void setLanguage(String value) => _language = value;
  String get baseUrl => _dio.options.baseUrl;

  Future<Map<String, dynamic>> get(String path, {Map<String, dynamic>? query}) async {
    final started = DateTime.now();
    final sw = Stopwatch()..start();
    try {
      final response = await _dio.get<Map<String, dynamic>>(path, queryParameters: query);
      final status = response.statusCode ?? 0;
      if (status >= 400) {
        throw DioException(requestOptions: response.requestOptions, response: response, type: DioExceptionType.badResponse);
      }
      final body = response.data ?? const {};
      _log(started, sw, 'GET', path, query, status: status, body: body);
      return body;
    } on DioException catch (error) {
      final mapped = _map(error);
      _log(started, sw, 'GET', path, query, status: error.response?.statusCode, error: mapped.toString());
      throw mapped;
    }
  }

  Future<Map<String, dynamic>> post(String path, {Object? data}) async {
    final started = DateTime.now();
    final sw = Stopwatch()..start();
    try {
      final response = await _dio.post<Map<String, dynamic>>(path, data: data);
      final status = response.statusCode ?? 0;
      if (status >= 400) {
        throw DioException(requestOptions: response.requestOptions, response: response, type: DioExceptionType.badResponse);
      }
      final body = response.data ?? const {};
      _log(started, sw, 'POST', path, const {}, status: status, body: body);
      return body;
    } on DioException catch (error) {
      final mapped = _map(error);
      _log(started, sw, 'POST', path, const {}, status: error.response?.statusCode, error: mapped.toString());
      throw mapped;
    }
  }

  void _log(DateTime started, Stopwatch sw, String method, String path, Map<String, dynamic>? query,
      {int? status, Map<String, dynamic>? body, String? error}) {
    if (RequestLog.active == null) return;
    String? summary;
    if (body != null) {
      final parts = <String>[];
      final prov = body['provenance'];
      final source = body['selected_source'] ?? (prov is Map ? prov['selected_source'] : null) ?? body['source'];
      if (source != null) parts.add('source=$source');
      final run = prov is Map ? prov['run_id'] : body['run_id'];
      if (run != null) parts.add('run=$run');
      if (body['status'] != null) parts.add('status=${body['status']}');
      if (body['hourly'] is List) parts.add('hourly=${(body['hourly'] as List).length}');
      if (body['daily'] is List) parts.add('daily=${(body['daily'] as List).length}');
      if (body['forecast'] is List) parts.add('forecast=${(body['forecast'] as List).length}');
      if (body['degraded'] == true) parts.add('DEGRADED');
      summary = parts.isEmpty ? null : parts.join(' · ');
    }
    RequestLog.record(RequestLogEntry(
      startedAt: started,
      method: method,
      path: path,
      query: query ?? const {},
      statusCode: status,
      durationMs: sw.elapsedMilliseconds,
      error: error,
      summary: summary,
    ));
  }

  AppApiError _map(DioException error) {
    if (error.type == DioExceptionType.receiveTimeout) {
      return const NetworkError('WeatherGPT is taking longer than usual. Please try again in a moment.');
    }
    if (error.type == DioExceptionType.connectionTimeout ||
        error.type == DioExceptionType.connectionError) {
      return const NetworkError('Could not reach WeatherGPT. Check your connection and try again.');
    }
    final status = error.response?.statusCode ?? 0;
    if (status == 400 || status == 422) return ValidationError(_detail(error) ?? 'Please check the requested weather data.');
    return ServerError(_detail(error) ?? 'WeatherGPT is temporarily unavailable. Please try again.');
  }
  String? _detail(DioException error) {
    final data = error.response?.data;
    if (data is Map) {
      if (data['detail'] != null) return '${data['detail']}';
      if (data['message'] != null) return '${data['message']}';
    }
    return null;
  }
}

File: src/core/services/apiClient.ts
/// The sole HTTP entry point — port of `lib/core/services/api_client.dart`
/// (Dio → fetch). Keep API keys on the server, never in this app. Every call
/// is timed, summarised and pushed into the request ring buffer.

import Constants from "expo-constants";

import { resolveBackendUrl } from "../config/backendConfig";
import { NetworkError, ServerError, ValidationError } from "../errors/appErrors";
import { decodeJsonObject } from "../models/jsonValues";
import { recordRequest } from "./requestLog";

let language = "en";

export function setApiLanguage(value: string): void {
  language = value;
}

function envBackendUrl(): string | null {
  // Expo public env vars survive `expo export`; Constants mirrors them too.
  const fromProcess = typeof process !== "undefined" ? process.env?.EXPO_PUBLIC_BACKEND_URL : undefined;
  const fromConstants = (Constants.expoConfig?.extra as Record<string, unknown> | undefined)?.BACKEND_URL;
  return (typeof fromProcess === "string" ? fromProcess : null) ?? (typeof fromConstants === "string" ? fromConstants : null);
}

export function apiBaseUrl(): string {
  return resolveBackendUrl({
    defineUrl: envBackendUrl(),
    dotenvUrl: null,
  });
}

function summarizeBody(body: Record<string, unknown> | null): string | null {
  if (!body) return null;
  const parts: string[] = [];
  const prov = body.provenance;
  const provMap = prov !== null && typeof prov === "object" ? (prov as Record<string, unknown>) : null;
  const source = body.selected_source ?? provMap?.selected_source ?? body.source;
  if (source !== null && source !== undefined) parts.push(`source=${source}`);
  const run = provMap?.run_id ?? body.run_id;
  if (run !== null && run !== undefined) parts.push(`run=${run}`);
  if (body.status !== null && body.status !== undefined) parts.push(`status=${body.status}`);
  if (Array.isArray(body.hourly)) parts.push(`hourly=${body.hourly.length}`);
  if (Array.isArray(body.daily)) parts.push(`daily=${body.daily.length}`);
  if (Array.isArray(body.forecast)) parts.push(`forecast=${body.forecast.length}`);
  if (body.degraded === true) parts.push("DEGRADED");
  return parts.length === 0 ? null : parts.join(" · ");
}

function mapFetchError(error: unknown, status: number, bodyText: string | null): Error {
  if (error instanceof NetworkError || error instanceof ServerError || error instanceof ValidationError) {
    return error;
  }
  const detail = (() => {
    if (!bodyText) return null;
    const data = decodeJsonObject(bodyText);
    if (data["detail"] !== null && data["detail"] !== undefined) return String(data["detail"]);
    if (data["message"] !== null && data["message"] !== undefined) return String(data["message"]);
    return null;
  })();

  if (error instanceof Error && error.name === "AbortError") {
    return new NetworkError("WeatherGPT is taking longer than usual. Please try again in a moment.");
  }
  if (error instanceof TypeError) {
    return new NetworkError("Could not reach WeatherGPT. Check your connection and try again.");
  }
  if (status === 400 || status === 422) {
    return new ValidationError(detail ?? "Please check the requested weather data.");
  }
  return new ServerError(detail ?? "WeatherGPT is temporarily unavailable. Please try again.");
}

async function request(
  method: "GET" | "POST",
  path: string,
  opts?: { query?: Record<string, unknown>; data?: unknown },
): Promise<Record<string, unknown>> {
  const startedAt = new Date();
  const startedMs = Date.now();
  const query = method === "GET" ? opts?.query ?? {} : {};
  let status = 0;
  let bodyText: string | null = null;

  const url = new URL(apiBaseUrl() + path);
  for (const [key, value] of Object.entries(query)) {
    if (value !== null && value !== undefined) url.searchParams.set(key, String(value));
  }

  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 60_000);
    let response: Response;
    try {
      response = await fetch(url.toString(), {
        method,
        headers: {
          Accept: "application/json",
          "Accept-Language": language,
          ...(method === "POST" ? { "Content-Type": "application/json" } : {}),
        },
        body: method === "POST" ? JSON.stringify(opts?.data ?? {}) : undefined,
        signal: controller.signal,
      });
    } finally {
      clearTimeout(timeout);
    }
    status = response.status;
    bodyText = await response.text();
    if (status >= 500) {
      throw new ServerError("WeatherGPT is temporarily unavailable. Please try again.");
    }
    if (status >= 400) {
      throw mapFetchError(new Error(`HTTP ${status}`), status, bodyText);
    }
    const body = decodeJsonObject(bodyText);
    recordRequest({
      startedAt,
      method,
      path,
      query,
      statusCode: status,
      durationMs: Date.now() - startedMs,
      summary: summarizeBody(body),
    });
    return body;
  } catch (error) {
    const mapped = mapFetchError(error, status, bodyText);
    recordRequest({
      startedAt,
      method,
      path,
      query,
      statusCode: status,
      durationMs: Date.now() - startedMs,
      error: mapped.message,
    });
    throw mapped;
  }
}

export const ApiClient = {
  get: (path: string, query?: Record<string, unknown>) => request("GET", path, { query }),
  post: (path: string, data?: unknown) => request("POST", path, { data }),
  baseUrl: apiBaseUrl,
};
